lib/api/apiUtils/object/corsResponse.js

```diff
@@ -0,0 +1,130 @@
+    may contain wildcards
+* @param {string} value - value from CORS request
+* @return {boolean} - true/false
+*/
+function _matchesValue(allowedValue, value) {
+    const wildcardIndex = allowedValue.indexOf('*');
+    // If no wildcards, simply return whether strings are equal
+    if (wildcardIndex === -1) {
+        return allowedValue === value;
+    }
+    // Otherwise make sure substrings match expected substrings before
+    // and after the wildcard
+    const beginValue = allowedValue.substring(0, wildcardIndex);
+    const endValue = allowedValue.substring(wildcardIndex + 1);
+    return (value.startsWith(beginValue) && value.endsWith(endValue));
+}
+
+/** _matchesOneOf - check if a value matches any values of a CORS rule array
+* @param {string[]} allowedValues - array of values allowed in CORS rule
+* @param {string} value - value from CORS request
+* @return {boolean} - true/false
+*/
+function _matchesOneOf(allowedValues, value) {
+    return allowedValues.some(allowedValue =>
+        _matchesValue(allowedValue, value));
+}
+
+/** _headersMatchRule - check if headers match AllowedHeaders of rule
+* @param {string[]} headers - Access-Control-Allow-Headers values or any headers
+*   that are sent in a CORS request
+* @param {string[]} ruleAllowedHeaders - AllowedHeaders of a CORS rule
+* @return {boolean} - true/false
+*/
+function _headersMatchRule(headers, ruleAllowedHeaders) {
+    if (!ruleAllowedHeaders) {
+        return false;
+    }
+    // AllowedHeaders may have been stored with uppercase letters
+    // during putBucketCors; ignore case when searching for match
+    const allowedHeaders = ruleAllowedHeaders.map(allowedHeader =>
+        allowedHeader.toLowerCase());
+    if (!headers.every(header => _matchesOneOf(allowedHeaders, header))) {
+        return false;
+    }
+    return true;
+}
+
+/** _findCorsRule - Return first matching rule in cors rules that permits
+*   CORS request
+* @param {object[]} rules - array of rules
+* @param {string} [rules.id] - optional id to identify rule
+* @param {string[]} rules[].allowedMethods - methods allowed for CORS
+* @param {string[]} rules[].allowedOrigins - origins allowed for CORS
+* @param {string[]} [rules[].allowedHeaders] - headers allowed in an
+*   OPTIONS request via the Access-Control-Request-Headers header
+* @param {number} [rules[].maxAgeSeconds] - seconds browsers should cache
+*   OPTIONS response
+* @param {string[]} [rules[].exposeHeaders] - headers to expose to external
+*   applications
+* @param {string} origin - origin of CORS request
+* @param {string} method - Access-Control-Request-Method value or the method of
+    a CORS request
+* @param {string[]} [headers] - Access-Control-Request-Headers values or any
+    headers that are sent in a CORS request
+* @return {(null|object)} - matching rule if found; null if no match
+*/
+export function findCorsRule(rules, origin, method, headers) {
+    for (let i = 0; i < rules.length; i++) {
```

Comment:
thanks for the suggestion: https://github.com/scality/S3/pull/518/commits/bdb7774c28ad1940b94f08554859c3c611a74d30